pkg/webhooks/leaderworkerset_webhook.go

```diff
@@ -80,42 +79,50 @@ var _ webhook.CustomValidator = &LeaderWorkerSetWebhook{}
+	return warnings, allErrs.ToAggregate()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *LeaderWorkerSetWebhook) ValidateDelete(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func (r *LeaderWorkerSetWebhook) generalValidate(ctx context.Context, obj runtime.Object) (admission.Warnings, field.ErrorList) {
 	lws := obj.(*v1.LeaderWorkerSet)
-	var allErrs []error
+	specPath := field.NewPath("spec")
+
+	var allErrs field.ErrorList
 	// Ensure replicas and groups number are valid
 	if lws.Spec.Replicas != nil && *lws.Spec.Replicas < 0 {
-		allErrs = append(allErrs, fmt.Errorf("the Replicas %d is invalid", lws.Spec.Replicas))
+		allErrs = append(allErrs, field.Invalid(specPath.Child("replicas"), lws.Spec.Replicas, "replicas must greater or equal to 0"))
 	}
 	if *lws.Spec.LeaderWorkerTemplate.Size < 1 {
-		allErrs = append(allErrs, fmt.Errorf("the Size %d is invalid", lws.Spec.LeaderWorkerTemplate.Size))
+		allErrs = append(allErrs, field.Invalid(specPath.Child("leaderWorkerTemplate", "size"), lws.Spec.LeaderWorkerTemplate.Size, "size must greater than 1"))
 	}
 	if int64(*lws.Spec.Replicas)*int64(*lws.Spec.LeaderWorkerTemplate.Size) > math.MaxInt32 {
-		allErrs = append(allErrs, fmt.Errorf("the product of replicas and worker replicas must not exceed %d for lws '%s'", math.MaxInt32, lws.Name))
+		// allErrs = append(allErrs, fmt.Errorf("the product of replicas and worker replicas must not exceed %d for lws '%s'", math.MaxInt32, lws.Name))
+		allErrs = append(allErrs, field.Invalid(specPath.Child("replicas"), lws.Spec.Replicas, fmt.Sprintf("the product of replicas and worker replicas must not exceed %d", math.MaxInt32)))
 	}
 	if lws.Spec.LeaderWorkerTemplate.RestartPolicy != v1.DefaultRestartPolicy && lws.Spec.LeaderWorkerTemplate.RestartPolicy != v1.RecreateGroupOnPodRestart {
-		allErrs = append(allErrs, fmt.Errorf("invalid value for leaderworkertemplate.restartpolicy %s", lws.Spec.LeaderWorkerTemplate.RestartPolicy))
-	}
-	return nil, errors.Join(allErrs...)
-}
+		allErrs = append(allErrs, field.Invalid(specPath.Child("leaderWorkerTemplate", "restartPolicy"), lws.Spec.LeaderWorkerTemplate.RestartPolicy, "only support [Default, RecreateGroupOnPodRestart]"))
 
-// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
-func (r *LeaderWorkerSetWebhook) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
-	var allErrs []error
-	newLws := newObj.(*v1.LeaderWorkerSet)
-	oldLws := oldObj.(*v1.LeaderWorkerSet)
-	if newLws.Spec.Replicas != nil && *newLws.Spec.Replicas < 0 {
-		allErrs = append(allErrs, fmt.Errorf("the Replicas %d is invalid", newLws.Spec.Replicas))
 	}
-	newLwsClone := newLws.DeepCopy()
-	// Replicas can be mutated
-	newLwsClone.Spec.Replicas = oldLws.Spec.Replicas
-	if !apiequality.Semantic.DeepEqual(oldLws.Spec, newLwsClone.Spec) {
```

Comment:
Why are we removing this logic? it might be better to do the opposite, by default a field is immutable unless we explicitly make it not.

Reply:
Is RolloutStrategy mutable in statefulsets?

Reply:
yes, we apply the same strategy.